test/e2e/scheduling/preemption.go

```diff
@@ -327,94 +328,91 @@ var _ = SIGDescribe("PreemptionExecutionPath", func() {
 					Resources: &v1.ResourceRequirements{
-						Requests: v1.ResourceList{fakecpu: resource.MustParse("50")},
-						Limits:   v1.ResourceList{fakecpu: resource.MustParse("50")},
+						Requests: v1.ResourceList{fakecpu: resource.MustParse("300")},
+						Limits:   v1.ResourceList{fakecpu: resource.MustParse("300")},
 					},
 				},
 			},
 			{
-				Replicas: int32(4),
+				Replicas: int32(1),
 				PodConfig: pausePodConfig{
 					Name:              "pod3",
 					Namespace:         ns,
 					Labels:            map[string]string{"name": "pod3"},
 					PriorityClassName: "p3",
 					NodeSelector:      map[string]string{"kubernetes.io/hostname": nodeHostNameLabel},
 					Resources: &v1.ResourceRequirements{
-						Requests: v1.ResourceList{fakecpu: resource.MustParse("95")},
-						Limits:   v1.ResourceList{fakecpu: resource.MustParse("95")},
-					},
-				},
-			},
-			{
-				Replicas: int32(1),
-				PodConfig: pausePodConfig{
-					Name:              "pod4",
-					Namespace:         ns,
-					Labels:            map[string]string{"name": "pod4"},
-					PriorityClassName: "p4",
-					NodeSelector:      map[string]string{"kubernetes.io/hostname": nodeHostNameLabel},
-					Resources: &v1.ResourceRequirements{
-						Requests: v1.ResourceList{fakecpu: resource.MustParse("400")},
-						Limits:   v1.ResourceList{fakecpu: resource.MustParse("400")},
+						Requests: v1.ResourceList{fakecpu: resource.MustParse("450")},
+						Limits:   v1.ResourceList{fakecpu: resource.MustParse("450")},
 					},
 				},
 			},
 		}
-		// create ReplicaSet{1,2,3} so as to occupy 780/800 fake resource
-		rsNum := len(rsConfs)
-		for i := 0; i < rsNum-1; i++ {
+		// create ReplicaSet{1,2,3} so as to occupy 950/1000 fake resource
+		for i := range rsConfs {
 			runPauseRS(f, rsConfs[i])
 		}
 
 		framework.Logf("pods created so far: %v", podNamesSeen)
 		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
 
-		// create ReplicaSet4
-		// if runPauseRS failed, it means ReplicaSet4 cannot be scheduled even after 1 minute
-		// which is unacceptable
-		runPauseRS(f, rsConfs[rsNum-1])
+		// create a Preemptor Pod
+		preemptorPodConf := pausePodConfig{
+			Name:              "pod4",
+			Namespace:         ns,
+			Labels:            map[string]string{"name": "pod4"},
+			PriorityClassName: "p4",
+			NodeSelector:      map[string]string{"kubernetes.io/hostname": nodeHostNameLabel},
+			Resources: &v1.ResourceRequirements{
+				Requests: v1.ResourceList{fakecpu: resource.MustParse("500")},
+				Limits:   v1.ResourceList{fakecpu: resource.MustParse("500")},
+			},
+		}
+		preemptorPod := createPod(f, preemptorPodConf)
+		waitForPreemptingWithTimeout(f, preemptorPod, framework.PodGetTimeout)
 
 		framework.Logf("pods created so far: %v", podNamesSeen)
 		framework.Logf("length of pods created so far: %v", len(podNamesSeen))
 
+		time.Sleep(10 * time.Second)
+		close(podNamesSeen)
```

Comment:
why close this? that creates a panic in the AddFunc

Reply:
If not closing it, the `for range podNamesSeen` loop will hang there waiting for the close.

It's like M producers (AddFunc) generating values, and one receiver says "ok, stop sending me more values". One solution is to close both controller and podNamesSeen here: (the `defer close(stopCh)` should also be adjusted):

```go
close(stopCh)
close(podNamesSeen)
```

Or the other option is to put the`for range podNamesSeen` loop into a goroutine:

```go
go podController.Run(stopCh)
defer func() {
	close(stopCh)
	close(podNamesSeen)
}()

...

maxRSPodsSeen := []int{1 * 2, 1 * 2, 1}
rsPodsSeen := []int{0, 0, 0}
go func() {
	for podName := range podNamesSeen {
		...
	}
}()

time.Sleep(10 * time.Second)
for i, got := range rsPodsSeen {
	...
}
```

WDTY?

Reply:
how about:

1. make individual vars to hold counts per rs:
```
var rs1, rs2, rs3 int32
```

2. increment those directly in AddFunc
```
if strings.HasPrefix(podName, "rs-pod1") {
	atomic.AddInt32(&rs1, 1)
} else if strings.HasPrefix(podName, "rs-pod2") {
	atomic.AddInt32(&rs2, 1)
} else if strings.HasPrefix(podName, "rs-pod3") {
	atomic.AddInt32(&rs3, 1)
}
```

3. after doing the preemption and waiting for the preempting pod to be scheduled, wait until the counts match our expectations:

```
err := wait.Poll(time.Second, wait.TestForeverTimeout, func() (bool, error) {
  if rs1 < 2 {
    framework.Logf("waiting for rs1 to observe 2 pod creations, got %d", rs1)
    return false, nil
  } else if rs1 > 2 {
    return false, fmt.Errorf("rs1 had more than 2 pods created: %d", rs1)
  }

  if rs2 < 2 {
    framework.Logf("waiting for rs2 to observe 2 pod creations, got %d", rs2)
    return false, nil
  } else if rs2 > 2 {
    return false, fmt.Errorf("rs2 had more than 2 pods created: %d", rs2)
  }

  if rs3 < 1 {
    framework.Logf("waiting for rs3 to observe 1 pod creations, got %d", rs3)
    return false, nil
  } else if rs3 > 1 {
    return false, fmt.Errorf("rs3 had more than 1 pods created: %d", rs3)
  }
})
if err != nil {
  framework.Failf("failed pod observation expectations: %v", err)
}
```

Reply:
(3) might be problematic: suppose at time T1, all the conditions are satisfied, then we return and think it's a success. However, the chances are: at a later time T2, the backfilled pods (rs1, rs2, rs3) might be over-preempted, i.e. the pods number may exceed the expected ones.

The other way is to check only once after a fixed interval (10 seconds let's say). But either way is not perfect.

Reply:
after 3 succeeds, waiting a fixed period (e.g. 5 seconds), then doing the check again to make sure nothing changed sounds fine

Reply:
SGTM will update. Thanks @liggitt.

Reply:
Updated. PTAL.